server/src/main/java/org/elasticsearch/action/support/nodes/TransportNodesAction.java

```diff
@@ -123,40 +126,90 @@ protected TransportNodesAction(
 
     @Override
     protected void doExecute(Task task, NodesRequest request, ActionListener<NodesResponse> listener) {
-        new AsyncAction(task, request, listener).start();
-    }
+        if (request.concreteNodes() == null) {
+            resolveRequest(request, clusterService.state());
+            assert request.concreteNodes() != null;
+        }
 
-    /**
-     * Map the responses into {@code nodeResponseClass} responses and {@link FailedNodeException}s, convert to a {@link NodesResponse} and
-     * pass it to the listener. Fails the listener with a {@link NullPointerException} if {@code nodesResponses} is null.
-     *
-     * @param request The associated request.
-     * @param nodeResponseTracker All node-level responses collected so far
-     * @throws NodeResponseTracker.DiscardedResponsesException if {@code nodeResponseTracker} has already discarded the intermediate results
-     * @see #newResponseAsync(Task, BaseNodesRequest, List, List, ActionListener)
-     */
-    // exposed for tests
-    void newResponse(Task task, NodesRequest request, NodeResponseTracker nodeResponseTracker, ActionListener<NodesResponse> listener)
-        throws NodeResponseTracker.DiscardedResponsesException {
+        final var mutex = new Object();
```

Comment:
I am doubting if we need the extra mutex here. I understand accessing the `responses` and to the `exceptions` individually needs to be synchronized but does accessing either of them need to be synchronized too?
If what I am saying above is not clear, maybe the following example will explain better what I have in mind. I am wondering if using the respective list as a lock is enough. For example:

```
              final ActionListener<NodeResponse> nodeResponseListener = ActionListener.notifyOnce(new ActionListener<>() {
                    @Override
                    public void onResponse(NodeResponse nodeResponse) {
                        synchronized (responses) {
                            responses.add(nodeResponse);
                        }
                    }

                    @Override
                    public void onFailure(Exception e) {
                        if (task instanceof CancellableTask cancellableTask && cancellableTask.isCancelled()) {
                            return;
                        }

                        logger.debug(() -> format("failed to execute [%s] on node [%s]", actionName, node), e);
                        synchronized (exceptions) {
                            exceptions.add(new FailedNodeException(node.getId(), "Failed node [" + node.getId() + "]", e));
                        }
                    }
```

What are your thoughts on this? Am I missing something?